ledger/voters.go

```diff
@@ -192,18 +196,14 @@ func (vt *votersTracker) newBlock(hdr bookkeeping.BlockHeader) {
 // Since the map is small (Usually  0 - 2 elements and up to StateProofMaxRecoveryIntervals) we decided to keep the code simple
 // and check for deletion in every round.
 func (vt *votersTracker) removeOldVoters(hdr bookkeeping.BlockHeader) {
-	// we calculate the lowest round for recovery according to the newest round (might be different from the rounds on cache)
-	proto := config.Consensus[hdr.CurrentProtocol]
-	recentRoundOnRecoveryPeriod := basics.Round(uint64(hdr.Round) - uint64(hdr.Round)%proto.StateProofInterval)
-	oldestRoundOnRecoveryPeriod := recentRoundOnRecoveryPeriod.SubSaturate(basics.Round(proto.StateProofInterval * proto.StateProofMaxRecoveryIntervals))
+	lowestStateProofRound := stateproof.GetOldestExpectedStateProof(&hdr)
 
 	for r, tr := range vt.votersForRoundCache {
 		commitRound := r + basics.Round(tr.Proto.StateProofVotersLookback)
 		stateProofRound := commitRound + basics.Round(tr.Proto.StateProofInterval)
 
 		// we remove voters that are no longer needed (i.e StateProofNextRound is larger ) or older than the recover period
-		if stateProofRound < hdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound ||
-			stateProofRound <= oldestRoundOnRecoveryPeriod {
+		if stateProofRound < lowestStateProofRound {
```

Comment:
the only difference here: the old code has **less or equal than `oldestRoundOnRecoveryPeriod`** and the new code has **less than** `lowestStateProofRound` that is returned by `GetOldestExpectedStateProof`

Reply:
you are right. I think we had an off-by-one bug. This is the comment I would add:
```
	// In case state proof chain stalls the votersTracker's map should be bounded to StateProofMaxRecoveryIntervals + 2
	// + 1 - since votersForRoundCache needs to contain an entry for a future state proof
	// + 1 - since votersForRoundCache needs to contain an entry to verify the earliest state proof
	// in the recovery interval. i.e. it needs to have an entry for R-StateProofMaxRecoveryIntervals-StateProofInterval
	// to verify R-StateProofMaxRecoveryIntervals
```

Does this make sense? I tried to confirm on on the unit tests.

Reply:
I agree. Let's save we have rnd=21, interval=5, and recovery intervals=1, then GetOldestExpectedStateProof returns 15, and we indeed should not delete 15.